### PR TITLE
needs review: add turbolinks support.

### PR DESCRIPTION
**Pull request by ohadlevy**

### CODE REVIEW

**Comment by ohadlevy** [comment]
@abenari do you mind reviewing this? I've noticed that link_to :back needs special handling.

**Comment by abenari** [comment]
A lot of pages are broken.
Because the javascripts are only loaded once with turbolinks, we need to load all of them on initial load. That include our javascripts host_edit, loookup_keys, compute_resource etc. and all the external ones like ace, spice, novnc etc.
This should be ok, except the initial load that might get a little long, and I hope we don't have namespace issues such as two methods with the same name, or global variable with the same name.

**Comment by ohadlevy** [comment]
@abenari I've added asset change detection, this should fix the missing javascript files.
this has a side effect of reloading the page twice (one time for turbolinks to detect, and another one to redirect) maybe we should consider merging all JS where it has no / little impact.

I did not find similar case of css?

**Comment by ohadlevy** [comment]
the tests failure are resolved with https://github.com/rails/turbolinks/commit/8eaf3f802be0934beafe65e53dfca7b2aadac974, lets hope they will release a new gem soon.

**Comment by domcleal** [comment]
Impressive!  As discussed on IRC, I think we should include some JS files for common actions in the main page layout so pages like /hosts don't cause full reloads.

**Comment by ohadlevy** [comment]
What do you think about full page caching?
add to your application.js somewhere:

Turbolinks.enableTransitionCache();

**Comment by domcleal** [comment]
That's interesting, it feels elegant as a user when viewing lists, though it is a bit jarring (as turbolinks warns) when you use two-pane.. e.g. navigate away from a page with a two-pane view open and it remembers that momentarily when you return, before it disappears again.

Adding data-no-transition-cache to the two-pane-right works ok (e.g. inside right_pane_content), so then the page doesn't get cached at all while the two-pane is open.

**Comment by domcleal** [comment]
Don't forget to update foreman.spec + file a redmine.

**Comment by ohadlevy** [comment]
I'll wait for the new gem to be released...

File: app/helpers/layout_helper.rb [hunk 1]
```diff
@@ -21,7 +21,7 @@ def stylesheet(*args)
   end
 
   def javascript(*args)
-    content_for(:javascripts) { javascript_include_tag(*args) }
+    content_for(:javascripts) { javascript_include_tag(*args, "data-turbolinks-track" => true) }
```
Comment (domcleal) [inline app/helpers/layout_helper.rb]: splat followed by args in Ruby 1.8 doesn't work, you'll need to do something like `args.push("data-turbolinks-track" => true)` above

File: foreman.spec [hunk 2]
```diff
@@ -79,6 +79,7 @@ Requires: %{?scl_prefix}rubygem(bootstrap-sass) >= 3.0.3.0
 Requires: %{?scl_prefix}rubygem(bootstrap-sass) < 3.0.4
 Requires: %{?scl_prefix}rubygem(foreigner) >= 1.4.2
 Requires: %{?scl_prefix}rubygem(deep_cloneable)
+Requires: %{?scl_prefix}rubygem(turbolinks) >= 2.2.2
```
Comment (domcleal) [inline foreman.spec:82]: This needs adding as a BuildRequires too, plus jquery-turbolinks (should that be in assets only?)

**Comment by domcleal** [comment]
Coming back to this @ohadlevy?

**Comment by isratrade** [comment]
@ohadlevy, the `data-id` seem to be added without dfcf758. Where does it breakdown if this commit is removed?

**Comment by isratrade** [comment]
@ohadlevy, tests only fail for Ruby 1.8.7 on PR #1476 which just removes the last commit dfcf758.

**Comment by domcleal** [comment]
Closing in favour of @isratrade's new PR.

**Pull request closed without merging**